python/ray/dashboard/modules/virtual_cluster/tests/test_virtual_cluster.py

```diff
@@ -0,0 +1,405 @@
+        revision=revision,
+    )
+    assert result["result"] is False
+    assert str(result["msg"]).endswith("No enough node instances to assign.")
+
+    # Create a new virtual cluster that the remaining nodes in the primary cluster
+    # are not enough.
+    result = create_or_update_virtual_cluster(
+        webui_url=webui_url,
+        virtual_cluster_id="virtual_cluster_2",
+        allocation_mode=allocation_mode,
+        replica_sets={"4c8g": 1, "8c16g": 1},
+        revision=0,
+    )
+    assert result["result"] is False
+    assert str(result["msg"]).endswith("No enough node instances to assign.")
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster_head",
+    [
+        {
+            "include_dashboard": True,
+        }
+    ],
+    indirect=True,
+)
+def test_remove_virtual_cluster(disable_aiohttp_cache, ray_start_cluster_head):
+    cluster: Cluster = ray_start_cluster_head
+    assert wait_until_server_available(cluster.webui_url) is True
+    webui_url = cluster.webui_url
+    webui_url = format_web_url(webui_url)
+
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "4c8g"})
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "8c16g"})
+
+    # Create a new virtual cluster with exclusive allocation mode.
+    result = create_or_update_virtual_cluster(
+        webui_url=webui_url,
+        virtual_cluster_id="virtual_cluster_1",
+        allocation_mode="exclusive",
+        replica_sets={"4c8g": 1, "8c16g": 1},
+        revision=0,
+    )
+    assert result["result"] is True
+
+    # Try removing a non-exist virtual cluster.
+    result = remove_virtual_cluster(
+        webui_url=webui_url, virtual_cluster_id="virtual_cluster_2"
+    )
+    assert result["result"] is False
+    # The error msg should tell us the virtual cluster does not exit.
+    assert str(result["msg"]).endswith("does not exist.")
+
+    # Remove the virtual cluster.
+    result = remove_virtual_cluster(
+        webui_url=webui_url, virtual_cluster_id="virtual_cluster_1"
+    )
+    assert result["result"] is True
+
+    # Create a new virtual cluster with mixed mode.
+    result = create_or_update_virtual_cluster(
+        webui_url=webui_url,
+        virtual_cluster_id="virtual_cluster_1",
+        allocation_mode="mixed",
+        replica_sets={"4c8g": 1, "8c16g": 1},
+        revision=0,
+    )
+    assert result["result"] is True
+
+    @ray.remote
+    class SmallActor:
+        def pid(self):
+            return os.getpid()
+
+    # Create an actor that requires some resources.
+    actor = SmallActor.options(num_cpus=0.1).remote()
+    ray.get(actor.pid.remote(), timeout=10)
+
+    # Remove the virtual cluster.
+    result = remove_virtual_cluster(
+        webui_url=webui_url, virtual_cluster_id="virtual_cluster_1"
+    )
+    # The virtual cluster can not be removed because some nodes
+    # are still in use.
+    assert result["result"] is False
+    assert str(result["msg"]).endswith("still in use.")
```

Comment:
We may have to tell users which nodes are still in use, so that they know which nodes to drain before removing. We should tell similar things when users are trying to scale down a virtual cluster.